lib/puppetserver/ca/generate_action.rb

```diff
@@ -0,0 +1,249 @@
+        cert
+      end
+
+      def extension_factory_for(ca, cert = nil)
+        ef = OpenSSL::X509::ExtensionFactory.new
+        ef.issuer_certificate  = ca
+        ef.subject_certificate = cert if cert
+
+        ef
+      end
+
+      def create_crl_for(ca_cert, ca_key, ca_ttl)
+        crl = OpenSSL::X509::CRL.new
+        crl.version = 1
+        crl.issuer = ca_cert.subject
+
+        ef = extension_factory_for(ca_cert)
+        crl.add_extension(
+          ef.create_extension(["authorityKeyIdentifier", "keyid:always", false]))
+        crl.add_extension(
+          OpenSSL::X509::Extension.new("crlNumber", OpenSSL::ASN1::Integer(0)))
+
+        crl.last_update = Time.now - 1
+        crl.next_update = Time.now + ca_ttl
+        crl.sign(ca_key, DEFAULT_SIGNING_DIGEST)
+
+        crl
+      end
+
+      def create_csr(key, name)
+        csr = OpenSSL::X509::Request.new
+
+        csr.public_key = key.public_key
+        csr.subject = OpenSSL::X509::Name.new([["CN", name]])
+        csr.version = 2
+        csr.sign(key, DEFAULT_SIGNING_DIGEST)
+
+        csr
+      end
+
+      def sign(ca_key, ca_cert, csr, ca_ttl)
+        cert = OpenSSL::X509::Certificate.new
+
+        cert.public_key = csr.public_key
+        cert.subject = csr.subject
+        cert.issuer = ca_cert.subject
+        cert.version = 2
+        cert.serial = 2
+
+        not_before = Time.now
+        cert.not_before = not_before
+        cert.not_after = not_before + ca_ttl
+
+        ef = extension_factory_for(ca_cert, cert)
+        CA_EXTENSIONS.each do |ext|
+          extension = ef.create_extension(*ext)
+          cert.add_extension(extension)
+        end
+
+        cert.sign(ca_key, DEFAULT_SIGNING_DIGEST)
+
+        cert
+      end
+
+      def parse(cli_args)
+        parser, inputs, unparsed = parse_inputs(cli_args)
+
+        if !unparsed.empty?
+          @logger.err 'Error:'
+          @logger.err 'Unknown arguments or flags:'
+          unparsed.each do |arg|
+            @logger.err "    #{arg}"
+          end
+
+          @logger.err ''
+          @logger.err parser.help
+
+          exit_code = 1
+        else
+          exit_code = nil
+        end
+
+        return inputs, exit_code
+      end
+
+      def parse_inputs(inputs)
+        parsed = {}
+        unparsed = []
+
+        parser = self.class.parser(parsed)
+
+        begin
+          parser.order!(inputs) do |nonopt|
+            unparsed << nonopt
+          end
+        rescue OptionParser::ParseError => e
+          unparsed += e.args
+          unparsed << inputs.shift unless inputs.first =~ /^-{1,2}/
+          retry
+        end
+
+        return parser, parsed, unparsed
+      end
+
+      def self.parser(parsed = {})
+        OptionParser.new do |opts|
+          opts.banner = BANNER
+          opts.on('--help', 'Display this generate specific help output') do |help|
+            parsed['help'] = true
+          end
+          opts.on('--version', 'Output the version') do |v|
```

Comment:
The fact that Import has a version flag is a bug. The main Cli class will parse version and return the gem's version. Individual actions don't have specific versions and we should remove this and the version option declaration in Import as well (I'm not asking for you to remove Import's as part of this PR though).

Reply:
Im remove it from generate and import as well.